<a href="https://colab.research.google.com/github/KasraRezaei/ML/blob/main/Fashion_MINST.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Fashion-MNIST contains:
60,000 training images
10,000 test images
Each image is 28×28 grayscale
10 clothing categories:
Label	Item
0	T-shirt/top
1	Trouser
2	Pullover
3	Dress
4	Coat
5	Sandal
6	Shirt
7	Sneaker
8	Bag
9	Ankle boot





In [45]:
from keras.models import Sequential
from keras.layers import Dense
from keras.datasets import fashion_mnist
from keras.utils import to_categorical
# Constants
RESHAPED = 784

# 10 type cloths
NB_CLASSES = 10
# Load Fashion-MNIST
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()
# Preprocess
x_train = x_train.reshape(-1, RESHAPED).astype("float32") / 255
x_test = x_test.reshape(-1, RESHAPED).astype("float32") / 255

y_train = to_categorical(y_train, NB_CLASSES)
# If label = 7 (Sneaker) >> [0,0,0,0,0,0,0,1,0,0]


# Shallow model >> one hidden layer
# 784 inputs → 64 neurons
# Input Parameters: 784 × 64 + 64 = 50,240 (Weights + biases)
# Output Parameters: 64 inputs → 10 outputs : 64 × 10(classes) + 10 = 650
y_test = to_categorical(y_test, NB_CLASSES)
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=(RESHAPED,)))
model.add(Dense(NB_CLASSES, activation='softmax'))
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

#Training
history = model.fit(
    x_train,
    y_train,
    epochs=10,
    batch_size=128,
    validation_split=0.2
)

#Evalouation:
score = model.evaluate(x_test, y_test)

# Get training accuracy from the last epoch
train_acc = history.history['accuracy'][-1]
# Calculate the gap
gap = train_acc - score[1]

print("Test score:", score[0])
print("Test accuracy:", score[1])
print("Gap (Training - Test):", gap)

Epoch 1/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.7264 - loss: 0.8399 - val_accuracy: 0.8365 - val_loss: 0.4736
Epoch 2/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.8472 - loss: 0.4510 - val_accuracy: 0.8529 - val_loss: 0.4271
Epoch 3/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.8588 - loss: 0.4075 - val_accuracy: 0.8527 - val_loss: 0.4206
Epoch 4/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8661 - loss: 0.3806 - val_accuracy: 0.8597 - val_loss: 0.3998
Epoch 5/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.8707 - loss: 0.3606 - val_accuracy: 0.8602 - val_loss: 0.3959
Epoch 6/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.8733 - loss: 0.3504 - val_accuracy: 0.8730 - val_loss: 0.3655
Epoch 7/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8822 - loss: 0.3300 - val_accuracy: 0.8732 - val_loss: 0.3588
Epoch 8/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.8832 - loss: 0.3223 - val_accuracy: 0.

Questions:
• What test accuracy does this shallow model achieve?

Part 2: Deepen the Network
Now make the network deeper by adding more hidden layers.

In [46]:
model = Sequential()

# Total parameters: Total ≈ 134,794 parameters
#  L1: 784 × 128 + 128 = 100,480
#  L2: 128 × 128 + 128 = 16,512
#  L3: 128 × 128 + 128 = 16,512
#  output: 128 × 10 + 10 = 1,290
model.add(Dense(128, activation='relu', input_shape=(RESHAPED,)))
model.add(Dense(128, activation='relu'))
model.add(Dense(128, activation='relu'))
model.add(Dense(NB_CLASSES, activation='softmax'))
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

#Training
history = model.fit(
    x_train,
    y_train,
    epochs=10,
    batch_size=128,
    validation_split=0.2
)

#Evalouation:
score = model.evaluate(x_test, y_test)

# Get training accuracy from the last epoch
train_acc = history.history['accuracy'][-1]
# Calculate the gap
gap = train_acc - score[1]

print("Test score:", score[0])
print("Test accuracy:", score[1])
print("Gap (Training - Test):", gap)

Epoch 1/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 8ms/step - accuracy: 0.7421 - loss: 0.7716 - val_accuracy: 0.8448 - val_loss: 0.4355
Epoch 2/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.8543 - loss: 0.4032 - val_accuracy: 0.8670 - val_loss: 0.3749
Epoch 3/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.8721 - loss: 0.3521 - val_accuracy: 0.8648 - val_loss: 0.3765
Epoch 4/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.8800 - loss: 0.3238 - val_accuracy: 0.8790 - val_loss: 0.3350
Epoch 5/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.8903 - loss: 0.2980 - val_accuracy: 0.8741 - val_loss: 0.3465
Epoch 6/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 12ms/step - accuracy: 0.8938 - loss: 0.2851 - val_accuracy: 0.8776 - val_loss: 0.3297
Epoch 7/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.8992 - loss: 0.2713 - val_accuracy: 0.8844 - val_loss: 0.3258
Epoch 8/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 8ms/step - accuracy: 0.9026 - loss: 0.2594 - val_accuracy: 0

Questions:
• How does performance change with added depth?
  Adding depth:
    - Increased training accuracy: ~88% → ~91%
    - Slightly increased validation/test accuracy ~87% → ~88%
    Training accuracy: 90.9%
    Test accuracy: 87.9%
    Gap ≈ 3%
    That’s mild overfitting beginning.
    The deeper model:
    Has more capacity
    Fits training data better
    Generalizes only slightly better


• Does training time increase?
  Yes — clearly, ~2–3 seconds per epoch
  Why?
    - First layer: 784 × 64 + 64 = 50,240
  New deep model (3 × 128 layers)
    Layer 1:
    784 × 128 + 128 = 100,480
    Layer 2:
    128 × 128 + 128 = 16,512
    Layer 3:
    128 × 128 + 128 = 16,512
    Output:
    128 × 10 + 10 = 1,290
    Total ≈ 134,794 parameters

Key Insight
Increasing depth:
✔ Improves training accuracy
✔ Slightly improves test accuracy
✔ Increases training time
✔ Increases overfitting risk
But improvement is diminishing.

Big Concept
Fully connected networks struggle with images because:
They treat images as flat vectors
They don’t preserve spatial locality
They can’t detect patterns like edges or textures explicitly
That’s why CNNs outperform them.

Why We Need Validation
Imagine this:
You train for many epochs.
If:
Training accuracy ↑
Validation accuracy ↓
That means:
The model is memorizing training data → overfitting.
Validation helps you decide:
When to stop training
Whether to change architecture
Whether to regularize


Part 3: Widen the Network
Now experiment with increasing the width (number of neurons per layer):

In [47]:
model = Sequential()
model.add(Dense(512, activation='relu', input_shape=(RESHAPED,)))
model.add(Dense(512, activation='relu'))
model.add(Dense(NB_CLASSES, activation='softmax'))
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

#Training
history = model.fit(
    x_train,
    y_train,
    epochs=10,
    batch_size=128,
    validation_split=0.2
)

#Evalouation:
score = model.evaluate(x_test, y_test)

# Get training accuracy from the last epoch
train_acc = history.history['accuracy'][-1]
# Calculate the gap
gap = train_acc - score[1]

print("Test score:", score[0])
print("Test accuracy:", score[1])
print("Gap (Training - Test):", gap)

Epoch 1/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 8s 18ms/step - accuracy: 0.7636 - loss: 0.6710 - val_accuracy: 0.8628 - val_loss: 0.3775
Epoch 2/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 8s 21ms/step - accuracy: 0.8647 - loss: 0.3716 - val_accuracy: 0.8733 - val_loss: 0.3543
Epoch 3/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 10s 21ms/step - accuracy: 0.8800 - loss: 0.3218 - val_accuracy: 0.8714 - val_loss: 0.3498
Epoch 4/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 7s 17ms/step - accuracy: 0.8893 - loss: 0.2944 - val_accuracy: 0.8829 - val_loss: 0.3257
Epoch 5/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 8s 20ms/step - accuracy: 0.8972 - loss: 0.2752 - val_accuracy: 0.8864 - val_loss: 0.3208
Epoch 6/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 8s 20ms/step - accuracy: 0.9052 - loss: 0.2574 - val_accuracy: 0.8825 - val_loss: 0.3289
Epoch 7/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 7s 18ms/step - accuracy: 0.9066 - loss: 0.2481 - val_accuracy: 0.8849 - val_loss: 0.3132
Epoch 8/10
375/375 ━━━━━━━━━━━━━━━━━━━━ 7s 20ms/step - accuracy: 0.9127 - loss: 0.2339 - val_acc

Answer:
Did Performance Improve?
Test accuracy before (128×3 model):
≈ 87.9%
Test accuracy now (512×2 model):
≈ 88.2%
🔎 Improvement:
Only about +0.3%
Very small.


Training Accuracy vs Test Accuracy
Final epoch:
Training accuracy ≈ 91.7%
Validation accuracy ≈ 88.8%
Test accuracy ≈ 88.2%
Gap:
Train ≈ 91.7%
Test  ≈ 88.2%
Gap   ≈ 3.5%
That gap is bigger than before.
👉 This means overfitting is increasing.
The model has more capacity and fits training data better, but generalization barely improves.

Did Training Time Increase?
Yes — significantly.
Previous model (128×3):
~3–5 seconds per epoch
New model (512×2):
~7–11 seconds per epoch
So training time roughly doubled or tripled.
Total:
≈ 669,706 parameters
That’s about 5× larger than your 128×3 model.

Important Insight:
You are now seeing a classic ML pattern:
Increasing model size always improves training accuracy
But test accuracy improves only up to a point
After that:
You just memorize better
You don’t generalize better




Questions
• How does widening compare to deepening?
Deepening (adding more hidden layers) slightly improves performance by allowing the network to learn more complex feature hierarchies.
Widening (adding more neurons per layer) increases the network’s capacity to store information but mostly helps with fitting training data rather than generalization.
• Does accuracy improve significantly?
Accuracy improves modestly with depth and width in fully connected networks on Fashion-MNIST.
Beyond a certain point (e.g., 512 neurons per layer), the improvement is very small (~0.3–1%) because the problem is relatively simple and fully connected layers don’t exploit spatial patterns.
• What about computational cost (training time, memory usage)?
Both widening and deepening increase computational cost.
Wider layers dramatically increase memory and multiply computations because the number of parameters grows quickly.
Deeper networks increase training time more gradually but add complexity.

Summary Reflections:
Increasing depth lets the model learn more complex representations, slightly improving generalization, but each added layer contributes diminishing returns. Increasing width allows the model to fit training data better, but generalization gains are limited, and computational cost rises sharply. The model stops improving when capacity exceeds what the data requires; at that point, accuracy saturates and overfitting may start, especially in fully connected networks. For Fashion-MNIST, fully connected layers reach ~88–89% test accuracy, and further depth or width brings little benefit.
